File: js/안전 영역(2468).js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');

const N = Number(input[0]);
// slice(1) - 첫 번째 줄을 제외한 나머지 줄 가져옴
const arr = input.slice(1).map(line => line.split(' ').map(Number));

// flat - 2차원 배열을 1차원 배열로 변환
const maxValue = Math.max(...arr.flat());

const answerArr = [];

for (let i=1; i < maxValue; i++) {
    // 얕은 복사 방법
    // const rainArr = [...arr];
    // 깊은 복사 방법
    // const rainArr = structuredClone(arr); -> 최신 브라우저에서만 지원
    const rainArr = arr.map(row => [...row]);
    for (let j=0; j < N; j++) {
        for (let k=0; k < N; k++) {
            if (arr[j][k] <= i) {
                rainArr[j][k] = 0;
            } else {
                rainArr[j][k] = 1;
            }
        }
    }

    const dx = [-1,1,0,0];
    const dy = [0,0,-1,1];

    function bfs(x,y) {
        const queue = [];
        queue.push([x,y]);
        // 방문 처리
        rainArr[x][y] = 2;

        while (queue.length) {
            const [x,y] = queue.shift();

            for (let t=0; t < 4; t++) {
                const nx = x + dx[t];
                const ny = y + dy[t];

                if (nx < 0 || nx >= N || ny < 0 || ny >= N) {
                    continue;
                }
                if (rainArr[nx][ny] === 0) {
                    continue;
                }
                if (rainArr[nx][ny] === 1) {
                    rainArr[nx][ny] = 2;
                    queue.push([nx,ny]);
                }
            }
        }
    }
    let cnt = 0;
    for (let a=0; a < N; a++) {
        for (let b=0; b < N; b++) {
            if (rainArr[a][b] == 1) {
                bfs(a,b);
                cnt++;
            }
        }
    }
    answerArr.push(cnt);
}

// Math.max(answerArr) -> NaN 반환 : 배열 자체를 넘기기 때문
// 따라서 전개연산자를 사용해서 배열 요소를 개별 인자로 펼침
// ex) answerArr = [1,2,3,4,5]
// Math.max(...answerArr) = Math.max(1,2,3,4,5)
console.log(answerArr.length === 0 ? 1 : Math.max(...answerArr));

File: python/안전 영역(2468).py
from collections import deque
import copy

N = int(input())
arr =[]
cnt_arr = []
# 최대 비 수량
max_rain = 0
for _ in range(N):
    a = list(map(int, input().split()))
    max_rain = max(max_rain, max(a))
    arr.append(a)

dx = [-1,1,0,0]
dy = [0,0,-1,1]

for i in range(1, max_rain):
    # 깊은 복사 방법 두 가지 
    # rain_arr = copy.deepcopy(arr)
    rain_arr = list(map(list, arr)) 
    for j in range(N):
        for k in range(N):
            if arr[j][k] <= i:
                rain_arr[j][k] = 0
            else:
                rain_arr[j][k] = 1
    
    def bfs(x,y):
        queue = deque()
        queue.append((x,y))
        # 방문 처리
        rain_arr[x][y] = 2

        while queue:
            x, y = queue.popleft()

            for i in range(4):
                nx = x + dx[i]
                ny = y + dy[i]

                if nx < 0 or nx >= N or ny < 0 or ny >= N:
                    continue
                if rain_arr[nx][ny] == 0:
                    continue
                if rain_arr[nx][ny] == 1:
                    rain_arr[nx][ny] = 2
                    queue.append((nx,ny))
    cnt = 0
    for a in range(N):
        for b in range(N):
            if rain_arr[a][b] == 1:
                bfs(a,b)
                cnt += 1
    cnt_arr.append(cnt)

print(max(cnt_arr) if cnt_arr else 1)